opt/scopeserver/dataserver/utils/loom_file_handler.py

```diff
@@ -80,71 +57,47 @@ def change_loom_mode(
             logger.info(f"{loom_file_path} now {self.active_looms[abs_file_path].get_connection().mode}")
 
-        logger.debug(f"Checking MD5")
-        new_partial_md5_hash = LoomFileHandler.get_partial_md5_hash(abs_file_path, 10000)
-        logger.debug(f"Old MD5 is: {partial_md5_hash} New MD5 is: {new_partial_md5_hash}")
-
-        if (partial_md5_hash != new_partial_md5_hash) and not keep_ss:
-            try:
-                logger.debug("Removing old SS")
-                os.remove(os.path.join(os.path.dirname(abs_file_path), partial_md5_hash + ".ss_pkl"))
-            except Exception as e:
-                logger.debug("Couldn't remove pickle SS")
-                logger.debug(e)
-                pass
         return self.active_looms[abs_file_path].get_connection()
 
-    def get_loom_absolute_file_path(self, loom_file_path: str) -> str:
-        return os.path.join(self.loom_dir, loom_file_path)
+    def get_loom_absolute_file_path(self, loom_file_path: Path) -> Path:
+        return self.loom_dir / loom_file_path
 
     def get_global_looms(self) -> list:
         return self.global_looms
 
     def set_global_data(self) -> None:
         self.global_looms = [x for x in os.listdir(self.loom_dir) if not os.path.isdir(os.path.join(self.loom_dir, x))]
 
-    def get_loom_connection(self, loom_file_path: str, mode: str = "r"):
+    def get_loom_connection(self, loom_file_path: Path, mode: str = "r") -> LoomConnection:
         logger.debug(f"Getting connection for {loom_file_path} in mode {mode}")
         return self.get_loom(loom_file_path=loom_file_path, mode=mode).get_connection()
 
-    def get_loom(self, loom_file_path: str, mode: str = "r"):
+    def get_loom(self, loom_file_path: Path, mode: str = "r") -> Loom:
         abs_loom_file_path = self.get_loom_absolute_file_path(loom_file_path)
-        if not os.path.exists(abs_loom_file_path):
-            logger.error(f"The file {loom_file_path} does not exists.")
-            raise ValueError("The file located at " + abs_loom_file_path + " does not exist.")
-        # To check if the given file path is given specified url!
-        partial_md5_hash = LoomFileHandler.get_partial_md5_hash(abs_loom_file_path, 10000)
-        if abs_loom_file_path in self.active_looms:
-            logger.debug("Should be preloaded")
-            try:
-                logger.debug(
-                    f"Current mode: {self.active_looms[abs_loom_file_path].get_connection().mode}, wanted mode {mode}"
-                )
-
-                if self.active_looms[abs_loom_file_path].get_connection().mode == mode:
-                    loom = self.active_looms[abs_loom_file_path]
+        with self.file_locks[abs_loom_file_path]:
+            if not os.path.exists(abs_loom_file_path):
+                logger.error(f"The file {loom_file_path} does not exists.")
+                raise ValueError(f"The file located at {abs_loom_file_path} does not exist.")
```

Comment:
Shouldn't this check be outside of the context manager ? Doesn't make sense to make a lock on a file that doesn't exist

Reply:
I think the check should be after the file is locked. That way you make the operations _atomic_ and you avoid a potential data race:
1. Thread A checks `loom0` exists
1. Thread A locks `loom0 mutex`
1. Thread B checks `loom0` exists
1. Thread A deletes `loom0`
1. Thread A yields
1. Thread B locks `loom0 mutex`
1. OOPS! Thread B thinks `loom0` exists, but Thread A already deleted it